pro_tes/ga4gh/tes/task_runs.py

```diff
@@ -57,30 +58,39 @@ def __init__(self) -> None:
             self.foca_config.db.dbs["taskStore"].collections["tasks"].client
         )
         self.store_logs = self.foca_config.storeLogs["execution_trace"]
+        self.task_distributor = TaskDistributionMiddleware()
 
     def create_task(  # pylint: disable=too-many-statements,too-many-branches
         self, **kwargs
     ) -> Dict:
         """Start task.
 
         Args:
-            **kwargs: Additional keyword arguments passed along with
-                             request and start_time.
+            **kwargs: Additional keyword arguments passed along with request.
+
         Returns:
             Task identifier.
         """
+        start_time = datetime.now().strftime("%m-%d-%Y %H:%M:%S")
         payload: Dict = deepcopy(request.json)
         db_document: DbDocument = DbDocument()
 
         db_document.basic_auth = self.parse_basic_auth(request.authorization)
 
+        db_document.task_outgoing = TesTask(**payload)
+
+        # middleware is called after the task is created in the database
+        payload = self.task_distributor.modify_request(request=request).json
+
         tes_uri_list = deepcopy(payload["tes_uri"])
         del payload["tes_uri"]
 
-        db_document.task_outgoing = TesTask(**payload)
         db_document.task_incoming = TesTask(**payload)
```

Comment:
I think I'm really confused. Didn't we say that we do _not_ want to modify the incoming but only the outgoing request object? At least that is what makes sense. If we modify the incoming request, it's not the incoming request anymore. And the outgoing request is a copy of the incoming request that was modified by the middlware, and with a ranked list of TES endpoints attached. And it is also the outgoing request that should be updated according to what is happening on the remote service and what is, essentially, returned to the user when they ask for info on that task (except a few properties like the start and other times pertaining to the task itself (not for the executor).

This all looks like it is actually done the other way around (or maybe I'm being stupid).

If that is the case, please open an issue to fix in another PR.

Reply:
Actually, task_outgoing is the one that is untouched, and all the changes (middleware,  the final response etc) are done in task_incoming.

Reply:
Yes, but it doesn't make sense. Don't you agree? If we modify the incoming request, it's not the incoming request anymore. Perhaps instead of naming them "incoming" and "outgoing", it might be better to call one of them just `task` and the other one `task_original`. Then `task` is the working copy which gets modified and always keeps all the info just as we want to return it to the user.

Reply:
If you want to rename, i.e task_incoming as task_outgoing and vice versa we can do that.

Reply:
> 

Yes makes sense, I am creating an issue for that where we just need to rename it.

Reply:
Created issue #137